internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package luks2
+
+import (
+	"bytes"
+	"crypto"
+	_ "crypto/sha1"
+	_ "crypto/sha256"
+	_ "crypto/sha512"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"golang.org/x/sys/unix"
+	"golang.org/x/xerrors"
+)
+
+var (
+	dataDeviceFstat = unix.Fstat
+)
+
+func cryptsetupLockDir() string {
+	return filepath.Join(runDir, "cryptsetup")
+}
+
+var isBlockDevice = func(mode os.FileMode) bool {
+	return mode&os.ModeDevice > 0 && mode&os.ModeCharDevice == 0
+}
+
+// LockMode defines the locking mode for DecodeHdr.
+type LockMode int
+
+const (
+	LockModeBlocking LockMode = iota
+	LockModeNonBlocking
+)
+
+// acquireSharedLock acquires an advisory shared lock on the LUKS volume associated with the
+// specified path. The ath can either be a block device or file containing a LUKS2 volume with
+// an integral header, or a detached header file associated with a LUKS device.
+//
+// If the mode parameter is LockModeBlocking, this function will block until the lock can be
+// obtained. If the mode parameter is LockModeNonBlocking, a wrapped syscall.Errno error with
+// the value of syscall.EWOULDBLOCK will be returned if the lock can not be obtained.
+//
+// A shared lock is for read-only access. There can be multiple parallel shared lock holders.
+//
+// This function implements the locking logic implemented by libcryptsetup - see
+// lib/utils_device_locking.c from the cryptsetup source code.
+//
+// On success, a callback is returned which should be called to release the lock.
+func acquireSharedLock(path string, mode LockMode) (release func(), err error) {
+	// Initially open the device or file for reading
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot open device: %w", err)
+	}
+	defer f.Close()
+
+	how := unix.LOCK_SH
+	if mode == LockModeNonBlocking {
+		how |= unix.LOCK_NB
+	}
+
+	// Obtain information about the opened device or file
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot obtain file info: %w", err)
+	}
+
+	var lockPath string
+	var openFlags int
+
+	switch {
+	case isBlockDevice(fi.Mode()):
+		// Block devices use an advisory lock on a file in /run/cryptsetup. The lock file
```

Comment:
this is in the main comment but I think it wouldn't hurt to repeat and write this:

// For block devices libcryptsetup uses ...

Reply:
I've updated these comments now.